lib/fprotect/fprotect_rram.c

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdint.h>
+#include <string.h>
+#include <errno.h>
+#include <soc.h>
+#include <nrf_erratas.h>
+#include <hal/nrf_rramc.h>
+
+#define kB_Pos (10)
+#define REGION_SIZE ((RRAMC_REGION_CONFIG_SIZE_Msk >> RRAMC_REGION_CONFIG_SIZE_Pos) << kB_Pos)
+
+static void protect_boot_code(uint32_t size){
+
+	nrf_rramc_region_config_t config={
+		.address = 0,
+		.permissions = NRF_RRAMC_REGION_PERM_READ_MASK | NRF_RRAMC_REGION_PERM_EXECUTE_MASK,
+		.writeonce = true,
+		.lock = true,
+		.size_kb = size>>kB_Pos,
+	};
+
+	nrf_rramc_region_config_set(NRF_RRAMC, 3, &config);
+}
+static void protect_boot_code2(uint32_t start, uint32_t size){
+	nrf_rramc_region_config_t config={
+		.address = start,
+		.permissions = NRF_RRAMC_REGION_PERM_READ_MASK | NRF_RRAMC_REGION_PERM_EXECUTE_MASK,
+		.writeonce = true,
+		.lock = true,
+		.size_kb = size>>kB_Pos,
+	};
+
+	nrf_rramc_region_config_set(NRF_RRAMC, 4, &config);
+}
+
+int fprotect_area(uint32_t start, size_t length)
+{
+	if (length <= REGION_SIZE) {
+		/*
+		 * Area fits in single region. Using the one with arbitrary start address.
+		 */
+		protect_boot_code2(start, length);
+
+	}else{
+		if (length > (REGION_SIZE + REGION_SIZE)) {
+			/*
+			 * Area larger than both regions combined.
+			 */
+			return -EINVAL;
+		}
+		if(start != 0){
+			/*
+			 * Region[3] has fixed 0x00000000 start address.
+			 * Using both regions implies this constraint.
+			 */
+			return -EINVAL;
+		}
+		protect_boot_code(REGION_SIZE);
+		protect_boot_code2(REGION_SIZE, length - REGION_SIZE);
```

Comment:
This part seems invalid. The run-time code should modify only the region 4. Region 3 should be configured by hardware based on the Immutable Bootloader configuration in UICR.
```suggestion
		return -EINVAL;
```